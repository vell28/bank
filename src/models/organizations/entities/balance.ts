import { CurrencyCodeType } from 'modules/currencies';

export interface IBalance {
  currency: CurrencyCodeType;
  balance: number;
  balanceHold: number;
  availableBalance: number;
}

export interface IAccountBalances {
  [key: string]: IBalance[];
}
