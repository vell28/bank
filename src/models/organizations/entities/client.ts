import { IAddress } from '../../operations/order-card';

export interface IClient {
  id?: string;
  email?: string;
  name?: string;
  phoneNumber?: string;
  citizenship?: string;
  address?: IAddress;
}
