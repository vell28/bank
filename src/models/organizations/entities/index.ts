export * from './account';
export * from './balance';
export * from './client';
export * from './organization';
