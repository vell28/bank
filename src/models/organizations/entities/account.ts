import { schema } from 'normalizr';

import { CurrencyCodeType } from 'modules/currencies';
import { IBalance } from './balance';

export enum AccountStatus {
  ACTIVE = 'ACTIVE',
  CLOSED = 'CLOSED',
  CLOSED_BY_CUSTOMER = 'CLOSED_BY_CUSTOMER',
  LOST = 'LOST',
  LOCKED = 'LOCKED',
}

export type AccountStatusType =
  | AccountStatus.ACTIVE
  | AccountStatus.CLOSED
  | AccountStatus.CLOSED_BY_CUSTOMER
  | AccountStatus.LOCKED
  | AccountStatus.LOST;

export enum AccountColors {
  DEFAULT = 'DEFAULT',
}

enum AvailableOperation {
  PAYMENT_SEPA = 'PAYMENT_SEPA',
  PAYMENT_SWIFT = 'PAYMENT_SWIFT',
  PAYMENT_CARD = 'PAYMENT_CARD',
  PAYMENT_CONTACT = 'PAYMENT_CONTACT',
  TOPUP_BY_CARD = 'TOPUP_BY_CARD',
  TOPUP_BY_CODE = 'TOPUP_BY_CODE',
  MONEY_REQUEST = 'MONEY_REQUEST',
  TOPUP_BY_TRANSFER = 'TOPUP_BY_TRANSFER',
}

type AvailableOperationType =
  | AvailableOperation.PAYMENT_SEPA
  | AvailableOperation.PAYMENT_SWIFT
  | AvailableOperation.PAYMENT_CONTACT
  | AvailableOperation.PAYMENT_CARD
  | AvailableOperation.MONEY_REQUEST
  | AvailableOperation.TOPUP_BY_CARD
  | AvailableOperation.TOPUP_BY_CODE
  | AvailableOperation.TOPUP_BY_TRANSFER;

export enum AccountTypes {
  PHYSICAL = 'PHYSICAL',
  JURIDICAL = 'JURIDICAL',
}

export type AccountType = AccountTypes.JURIDICAL | AccountTypes.PHYSICAL;

export interface IAccount {
  id: string;
  accountType: AccountType;
  number: string;
  title?: string;
  status: AccountStatusType;
  currency: {
    code: CurrencyCodeType;
  };
  clientId: string;
  color: AccountColors.DEFAULT;
  accountName: string;
  bonusProgramActivate: boolean;
  balance: number;
  balanceHold: number;
  accruedInterest: number | null;
  availableBalance: number;
  balances: IBalance[];
  createdAt: string;
  availableOperations: AvailableOperationType[];
}

export interface IUpdateBalanceResponse {
  balances: IBalance[];
}

export interface IAccountsNormalize {
  [key: string]: IAccount;
}

export const accountSchema = new schema.Entity('account', {}, { idAttribute: 'id' });
export const accountListSchema = new schema.Array(accountSchema);
