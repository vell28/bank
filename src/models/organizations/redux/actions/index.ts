import { Dispatch } from 'redux';

import { fetchOrganizationRequest, fetchAccountBalanceRequest } from 'modules/api-requests/organizations';
import { getStateType } from 'modules/store/types';
import { getAccountsIds } from '../selectors';
import {
  FETCH_ORGANIZATIONS,
  IFetchOrganizationsAction,
  IUpdateAccountBalanceAction,
  UPDATE_ACCOUNT_BALANCE,
} from './entities';

export const fetchOrganizations = (): IFetchOrganizationsAction => ({
  type: FETCH_ORGANIZATIONS,
  payload: fetchOrganizationRequest(),
});

export const updateAccountBalance = (id: string): IUpdateAccountBalanceAction => ({
  type: UPDATE_ACCOUNT_BALANCE,
  payload: fetchAccountBalanceRequest(id),
  meta: id,
});

export const updateAllAccountBalance = (dispatch: Dispatch, getState: getStateType) => {
  const ids = getAccountsIds(getState());
  const promises = ids.map((id: string) => dispatch(updateAccountBalance(id)));
  return Promise.all(promises);
};
