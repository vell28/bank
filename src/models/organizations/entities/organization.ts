import { IClient } from './client';
import { ICard } from '../../operations/card-settings/entities';
import { IAccount } from './account';

export interface IOrganization {
  id: number;
  name: string;
  title: string;
  tin: string;
  clientName: string;
  client: IClient;
  accounts: IAccount[];
  cards: ICard[];
}
