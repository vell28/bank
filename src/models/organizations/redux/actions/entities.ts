import { IAsyncAction } from 'modules/store/types';
import { ICard } from 'models/operations/card-settings/entities';
import { IUpdateBalanceResponse, IAccount } from '../../entities';

export const FETCH_ORGANIZATIONS = 'organizations/FETCH';
export const UPDATE_ACCOUNT_BALANCE = 'organizations/UPDATE_ACCOUNT_BALANCE';

export interface IUpdateAccountBalanceAction extends IAsyncAction<IUpdateBalanceResponse> {
  meta: string; // account id
}

// this created due to incompability of adresses that server send and consume
// shouldn't be used anywhere except map serverTypes to clientTypes function
export interface IServerAddress {
  apartmentNumber?: string;
  city?: string;
  country?: string;
  postalCode?: string;
  province?: string;
  street?: string;
  streetNumber?: string;
}
export interface IServerClient {
  id?: string;
  email?: string;
  name?: string;
  phoneNumber?: string;
  citizenship?: string;
  address?: IServerAddress;
}

export interface IServerOrganization {
  id: number;
  name: string;
  title: string;
  tin: string;
  clientName: string;
  client: IServerClient;
  accounts: IAccount[];
  cards: ICard[];
}

export type IFetchOrganizationsAction = IAsyncAction<IServerOrganization[]>;
