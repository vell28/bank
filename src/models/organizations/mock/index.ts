import { CurrencyCodes } from 'modules/currencies';
import {
  IAccount, AccountStatus, AccountTypes, AccountColors
} from '../entities';
import { ICard, CardProduct, Limit } from '../../operations/card-settings/entities';

import { IServerClient } from '../redux/actions/entities';
import { exampleServerAddress } from '../../operations/order-card/mock';

export const cardData: ICard = {
  accountId: 'PPY5107',
  availableBalance: 1897.61,
  balance: 0,
  balanceHold: 0,
  balances: [
    {
      currency: CurrencyCodes.EUR,
      balance: 1030.05,
      balanceHold: 0,
      availableBalance: 1897.61,
    },
  ],
  clientId: '6449',
  currency: { code: CurrencyCodes.EUR },
  expireAt: '2022-05-01T00:00:00',
  id: 'PPY13643',
  limits: [
    {
      type: Limit.DAY,
      title: null,
      maxValue: 500,
      usedValue: 0,
    },
  ],
  number: '5277 60_3 984',
  options: {
    limits: { disable: false },
  },
  owner: { embossedName: 'RALF WILLIAMS' },
  paySystem: 'MASTERCARD',
  plasticStatus: 'PENDING_SHIPMENT',
  productType: CardProduct.MAIN,
  status: AccountStatus.ACTIVE,
};

export const accountsData: IAccount[] = [
  {
    accountName: 'MT98PAPY36836000002676370004519',
    accountType: AccountTypes.PHYSICAL,
    accruedInterest: null,
    availableBalance: 1897.61,
    availableOperations: [],
    balance: 0,
    balanceHold: 0,
    color: AccountColors.DEFAULT,
    balances: [
      {
        currency: CurrencyCodes.EUR,
        balance: 1030.05,
        balanceHold: 0,
        availableBalance: 1897.61,
      },
    ],
    bonusProgramActivate: false,
    clientId: '6449',
    createdAt: '2019-11-24T14:59:55',
    currency: { code: CurrencyCodes.EUR },
    id: 'PPY5107',
    number: 'MT98PAPY36836000002676370004519',
    status: AccountStatus.ACTIVE,
  },
  {
    accountName: 'MT98PAPY36836000002676370004519',
    accountType: AccountTypes.JURIDICAL,
    accruedInterest: null,
    availableBalance: 1897.61,
    availableOperations: [],
    balance: 0,
    balanceHold: 0,
    color: AccountColors.DEFAULT,
    balances: [
      {
        currency: CurrencyCodes.EUR,
        balance: 1030.05,
        balanceHold: 0,
        availableBalance: 1897.61,
      },
    ],
    bonusProgramActivate: false,
    clientId: '6449',
    createdAt: '2019-11-24T14:59:55',
    currency: { code: CurrencyCodes.EUR },
    id: 'PPY6666',
    number: 'MT98PAPY36836000002676370004519',
    status: AccountStatus.ACTIVE,
  },
];

export const clientData: IServerClient = {
  id: '42',
  name: 'Test Client',
  citizenship: 'Germany',
  email: 'test.client@test.com',
  phoneNumber: '222333',
  address: exampleServerAddress,
};
