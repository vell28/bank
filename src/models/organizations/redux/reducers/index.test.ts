import { organizations, IOrganizationsState } from '.';
import { FETCH_ORGANIZATIONS, IServerOrganization } from '../actions/entities';
import { accountsData, clientData } from '../../mock';

const initAccounts: IOrganizationsState = {
  isLoading: false,
  isSuccess: false,
  isError: false,
  data: [],
  errors: [],
};

describe('organizations reducers', () => {
  test('organizations/FETCH_ACCOUNTS', () => {
    const organization: IServerOrganization = {
      id: 0,
      accounts: accountsData,
      client: clientData,
      name: 'testOrg',
      cards: [],
      clientName: 'Duplicate?',
      tin: '',
      title: 'title',
    };

    const data: IServerOrganization[] = [organization];

    const action = {
      type: `${FETCH_ORGANIZATIONS}_SUCCESS`,
      payload: {
        data,
      },
    };

    const newState = organizations(initAccounts, action);
    expect(newState.data).toMatchSnapshot();
  });
});
