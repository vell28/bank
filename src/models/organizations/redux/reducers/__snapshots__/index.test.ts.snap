// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organizations reducers organizations/FETCH_ACCOUNTS 1`] = `
Array [
  Object {
    "accounts": Array [
      Object {
        "accountName": "MT98PAPY36836000002676370004519",
        "accountType": "PHYSICAL",
        "accruedInterest": null,
        "availableBalance": 1897.61,
        "availableOperations": Array [],
        "balance": 0,
        "balanceHold": 0,
        "balances": Array [
          Object {
            "availableBalance": 1897.61,
            "balance": 1030.05,
            "balanceHold": 0,
            "currency": "EUR",
          },
        ],
        "bonusProgramActivate": false,
        "clientId": "6449",
        "color": "DEFAULT",
        "createdAt": "2019-11-24T14:59:55",
        "currency": Object {
          "code": "EUR",
        },
        "id": "PPY5107",
        "number": "MT98PAPY36836000002676370004519",
        "status": "ACTIVE",
      },
      Object {
        "accountName": "MT98PAPY36836000002676370004519",
        "accountType": "JURIDICAL",
        "accruedInterest": null,
        "availableBalance": 1897.61,
        "availableOperations": Array [],
        "balance": 0,
        "balanceHold": 0,
        "balances": Array [
          Object {
            "availableBalance": 1897.61,
            "balance": 1030.05,
            "balanceHold": 0,
            "currency": "EUR",
          },
        ],
        "bonusProgramActivate": false,
        "clientId": "6449",
        "color": "DEFAULT",
        "createdAt": "2019-11-24T14:59:55",
        "currency": Object {
          "code": "EUR",
        },
        "id": "PPY6666",
        "number": "MT98PAPY36836000002676370004519",
        "status": "ACTIVE",
      },
    ],
    "cards": Array [],
    "client": Object {
      "address": Object {
        "apartmentNumber": 42,
        "city": "Düsseldorf",
        "countryCode": "DE",
        "postalCode": "40479",
        "province": "North Rhine-Westphalia",
        "street": "Gartenstraße",
        "streetNumber": "53",
      },
      "citizenship": "Germany",
      "email": "test.client@test.com",
      "id": "42",
      "name": "Test Client",
      "phoneNumber": "222333",
    },
    "clientName": "Duplicate?",
    "id": 0,
    "name": "testOrg",
    "tin": "",
    "title": "title",
  },
]
`;
